### PR TITLE
Convert BooleanClause class to record class

### PR DESCRIPTION
**Pull request by Pulkitg64**
### Description

Convert ```BooleanClause``` class to record class which addresses #13207 partially.

### CODE REVIEW

**Review by jpountz — APPROVED** [review]
Please add a note to `lucene/MIGRATE.txt`, otherwise it looks good to me.

Do you have thoughts @uschindler ?

**Comment by jpountz** [comment]
> Please add a note to lucene/MIGRATE.txt

By the way, I wonder how we should handle all the classes that we are converting to records, should we have a single section in `lucene/MIGRATE.txt` that lists all classes for simplicity, instead of having one entry per class?

**Comment by Pulkitg64** [comment]
That's a good idea @jpountz. I can add this change under this section: https://github.com/apache/lucene/blob/main/lucene/MIGRATE.md#some-classes-converted-to-records-classes-github13207

**Comment by uschindler** [comment]
> That's a good idea @jpountz. I can add this change under this section: https://github.com/apache/lucene/blob/main/lucene/MIGRATE.md#some-classes-converted-to-records-classes-github13207

That was also my plan. This is why I created the section in the migration guide. No need for separate sections, but a new entry in the enumeration is required.

File: lucene/core/src/java/org/apache/lucene/search/BooleanClause.java [hunk 1]
```diff
@@ -67,25 +67,12 @@ public String toString() {
     };
   }
 
-  /** The query whose matching documents are combined by the boolean query. */
-  private final Query query;
-
-  private final Occur occur;
-
   /** Constructs a BooleanClause. */
   public BooleanClause(Query query, Occur occur) {
```
Comment (uschindler) [inline lucene/core/src/java/org/apache/lucene/search/BooleanClause.java:71]: Please replace this by the generic constructor without parameters and brackets.

`public BooleanClause {`

Just check parameters, no assignment.

**Review by uschindler — CHANGES_REQUESTED** [review]

File: lucene/core/src/java/org/apache/lucene/search/BooleanClause.java (hunk 1 above)
Comment (uschindler) [reply lucene/core/src/java/org/apache/lucene/search/BooleanClause.java:71]: Like this: https://github.com/apache/lucene/blob/54a2e11434d81ab92923bccbe651752f81ac1973/lucene/core/src/java/org/apache/lucene/store/IOContext.java#L61

Comment (Pulkitg64) [reply lucene/core/src/java/org/apache/lucene/search/BooleanClause.java:71]: Makes sense. On it.

Comment (Pulkitg64) [reply lucene/core/src/java/org/apache/lucene/search/BooleanClause.java]: Changed in the new commit.

**Review by uschindler — APPROVED** [review]

**Comment by uschindler** [comment]
The failing test is a known issue (see mailing list). Just ignore that.

**Comment by wardle** [comment]
Hi all. I understand the change to a record class but this does create a breaking change that would be avoided through the addition of deprecated accessors matching the old class ie getQuery and getOccur, which could be later removed in Lucene 11. Obviously other changes have been easier to manage in moving from Lucene 9->10 because methods have been flagged as deprecated in 9 and it has been straightforward to prepare for these changes. It is tricky because for users of Lucene with codebases that might run on either Java 11, or 21, one previously could have a codebase that would run using different Lucene versions, but this change means one has to choose the correct Lucene version AND keep two branches of their code.

**Comment by uschindler** [comment]
Hi,
This was planned to be backwards incompatible. There is a migration guide entry. 

Lucene generally revises APIs between major versions in incompatible ways. With Lucene 9 this was even more dramatic as a large part of the class and package structure was renamed to be Java module system compatible.

Lucene 10 was released last week and there won't be any changes anymore.

Uwe

**Pull request merged**